Quotient/events/roommessageevent.cpp

```diff
@@ -219,6 +219,49 @@ QString RoomMessageEvent::replacedBy() const
     // clang-format on
 }
 
+bool RoomMessageEvent::isReply() const
+{
+    const auto relations = contentPart<QJsonObject>("m.relates_to"_ls);
+    if (!relations.isEmpty()) {
+        const bool hasReplyRelation = relations.contains("m.in_reply_to"_ls);
+        bool isFallingBack = relations["is_falling_back"_ls].toBool();
+        return hasReplyRelation && !isFallingBack;
+    }
+    return false;
+}
+
+bool RoomMessageEvent::isReplyIncludingFallbacks() const
+{
+    return contentPart<QJsonObject>("m.relates_to"_ls).contains("m.in_reply_to"_ls);
+}
+
+QString RoomMessageEvent::replyEventId() const
+{
+    return contentJson()["m.relates_to"_ls].toObject()["m.in_reply_to"_ls].toObject()["event_id"_ls].toString();
+}
+
+bool RoomMessageEvent::isThreaded() const
+{
+    return (contentPart<QJsonObject>("m.relates_to"_ls).contains("rel_type"_ls)
+            && contentPart<QJsonObject>("m.relates_to"_ls)["rel_type"_ls].toString() == "m.thread"_ls)
+            || (!unsignedPart<QJsonObject>("m.relations"_ls).isEmpty() && unsignedPart<QJsonObject>("m.relations"_ls).contains("m.thread"_ls));
```

Comment:
Add `EventRelation::ThreadType` next to other types, then:
```suggestion
    const auto relation = getUpstreamRelation();
    return (relation && relation.type == EventRelation::ThreadType)
            || unsignedPart<QJsonObject>("m.relations"_ls).contains(EventRelation::ThreadType);
```
You are not winning anything if you check `m.relations` for emptiness first; `QJsonObject::contains()` will almost always catch that case and exit early.